modules/sheet/editTags.js

```diff
@@ -0,0 +1,137 @@
+        close()
+        onSave()
+      }
+    })
+
+    function edit ({ msgId }, cb) {
+      const tagsToCreate = MutantArray([])
+      const tagsToApply = MutantArray([])
+      const tagsToRemove = MutantArray([])
+      const tagsInput = Value('')
+
+      const myId = api.keys.sync.id()
+      const messageTags = map(ScuttleTag.obs.messageTagsFrom(msgId, myId), tagId =>
+        ScuttleTag.obs.Tag(tagId, api.about.obs.name))
+      const filteredMessages = computed([messageTags, tagsToRemove], (tags, removedIds) =>
+        filter(tags, tag => !removedIds.includes(tag.tagId)))
+
+      const messageTagsView = map(filteredMessages, tag =>
+        computed(tag, t => HtmlTag(t, () => tagsToRemove.push(t.tagId))))
+      const tagsToApplyView = map(tagsToApply, tag =>
+        HtmlTag(tag, () => tagsToApply.delete(tag)))
+      const tagsToCreateView = map(tagsToCreate, tag =>
+        HtmlTag({ tagName: tag, tagId: 'new' }, () => tagsToCreate.delete(tag)))
+      const stagedTags = computed([messageTagsView, tagsToApplyView, tagsToCreateView], (a, b, c) =>
+        h('StagedTags', concat(a, [b, c])))
+
+      const input = h('input.tags', {
+        placeholder: 'Add tags here',
+        'ev-keyup': onInput,
+        value: tagsInput()
+      })
+
+      input.addEventListener('suggestselect', onSuggestSelect)
+
+      return {
+        content: [stagedTags, h('EditTags', input)],
+        onMount,
+        onSave
+      }
+
+      function onMount () {
+        input.focus()
+        const appliedTagIds = map(filteredMessages, tag => tag.tagId)
+        const applyTagIds = map(tagsToApply, tag => tag.tagId)
+        const stagedTagIds = computed([ appliedTagIds, applyTagIds ], (a, b) => concat(a, b))
+        const getTagSuggestions = api.tag.obs.suggest(stagedTagIds)
+        addSuggest(input, (inputText, cb) => {
+          cb(null, getTagSuggestions(inputText))
+        }, { cls: 'SuggestBox' })
+      }
+
+      function onInput (e) {
+        const input = e.target.value
+        if (!input.endsWith(',')) {
+          tagsInput.set(input)
+          return
+        }
+        const tag = input.substring(0, input.length - 1)
+        tagsToCreate.push(tag)
+        e.target.value = ''
+      }
+
+      function onSuggestSelect (e) {
+        e.target.value = ''
+        const { value, tagId } = e.detail
+        if (!tagId) {
+          tagsToCreate.push(value)
+          return
+        }
+        const index = tagsToRemove().indexOf(tagId)
+        if (index >= 0) {
+          tagsToRemove.deleteAt(index)
+        } else {
+          tagsToApply.push({ tagId, tagName: value })
+        }
+      }
+
+      function onSave () {
+        // tagsToCreate
+        forEach(tagsToCreate(), tag => {
+          ScuttleTag.async.create(null, (err, msg) => {
+            if (err) return
+            ScuttleTag.async.name({ tag: msg.key, name: tag }, cb)
+            ScuttleTag.async.apply({ tagged: true, message: msgId, tag: msg.key }, cb)
```

Comment:
I'm not sure, but you perhaps don't want to call cb twice here?
Don't know what tend to be passed in here.
You could run these in parallel then run the cb once after they both finish with something like https://github.com/feross/run-parallel ... or just run them one after the other like nest callbacks

Reply:
@mixmix Yeah, I remember writing this and thinking that this callback was gonna get hit a ton. At the time I just wanted to get it working so it wasn't a huge issue.

Taking a look at it now, setting it up to that these all run and callback once should definitely be possible but could be a bit complex I think. `tagsToApply` and `tagsToRemove` aren't that bad, but `tagsToCreate` needs to first create all the tags, then take those tags and name/apply them.

I'm not used to coding with callbacks so while I'm sure its possible how to set it up is not immediately obvious to me. I'll revisit it soon hopefully, but if you have any more pointers that would be helpful as well.

Reply:
Finally got around to looking at this again, and seems to be working great with run-parallel, thanks for the pointer @mixmix 😄